spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/encryption/MongoClientEncryption.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.mongodb.core.encryption;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import org.bson.BsonBinary;
+import org.bson.BsonValue;
+import org.springframework.data.mongodb.core.encryption.EncryptionKey.Type;
+import org.springframework.util.Assert;
+
+import com.mongodb.client.model.vault.EncryptOptions;
+import com.mongodb.client.vault.ClientEncryption;
+
+/**
+ * {@link ClientEncryption} based {@link Encryption} implementation.
+ *
+ * @author Christoph Strobl
+ * @since 4.1
+ */
+public class MongoClientEncryption implements Encryption<BsonValue, BsonBinary> {
+
+	private final Supplier<ClientEncryption> source;
+	private final AtomicReference<ClientEncryption> cached;
+
+	private MongoClientEncryption(Supplier<ClientEncryption> source) {
+
+		this.source = source;
+		this.cached = new AtomicReference<>(source.get());
+	}
+
+	/**
+	 * The caching {@link MongoClientEncryption} variant caches and reuses the {@link ClientEncryption} obtained from the
+	 * {@link Supplier} until explicitly {@link #refresh() refreshed}.
+	 * 
+	 * @param clientEncryption must not be {@literal null} nor emit {@literal null}.
+	 * @return new instance of {@link MongoClientEncryption}.
+	 */
+	public static MongoClientEncryption caching(Supplier<ClientEncryption> clientEncryption) {
+		return new MongoClientEncryption(clientEncryption);
+	}
+
+	/**
+	 * Create a new {@link MongoClientEncryption} instance for the given {@link ClientEncryption}.
+	 *
+	 * @param clientEncryption must not be {@literal null}.
+	 * @return new instance of {@link MongoClientEncryption}.
+	 */
+	public static MongoClientEncryption just(ClientEncryption clientEncryption) {
+
+		Assert.notNull(clientEncryption, "ClientEncryption must not be null");
+
+		return new MongoClientEncryption(() -> clientEncryption) {
+			@Override
+			public boolean refresh() {
+				return false;
+			}
+		};
+	}
+
+	/**
+	 * @return {@literal true} if refreshed, {@literal false} otherwise.
+	 */
+	public boolean refresh() {
```

Comment:
I'm wondering whether we should rather show an `Lazy` example for singleton usage and always call `Supplier.get()`.

Reply:
I had thought about using `Lazy` but decided against it to have more control over the refresh. Also calling the supplier may lead to unnecessary creation of `ClientEncryption` instances that all hold connections to the database.